python/cudf/cudf/tests/test_spilling.py

```diff
@@ -594,3 +594,43 @@ def test_statistics_expose(manager: SpillManager):
     assert stat.count == 10
     assert stat.total_nbytes == buffers[0].nbytes * 10
     assert stat.spilled_nbytes == buffers[0].nbytes * 10
+
+
+def _statistics_rmm_fragmentation(size: int):
+    manager = get_global_manager()
+    b1 = as_buffer(data=rmm.DeviceBuffer(size=size), exposed=False)
+    b2 = as_buffer(data=rmm.DeviceBuffer(size=size), exposed=False)
+    b3 = as_buffer(data=rmm.DeviceBuffer(size=size), exposed=False)
+    assert manager.statistics.lowest_utilization is None
+
+    # Trigger spilling, no fragmentation
+    b4 = as_buffer(data=rmm.DeviceBuffer(size=size), exposed=False)
+    assert manager.statistics.lowest_utilization == size * 3
+
+    # By freeing the first and last buffer, we now have fragmentation:
+    # total free memory is `size * 2` but fragmented into two segments.
+    # The result is a "lowest utilization" of `size * 2`.
+    del b1, b4
+    as_buffer(data=rmm.DeviceBuffer(size=size * 2), exposed=False)
+    assert manager.statistics.lowest_utilization == size * 2
+    del b2, b3
```

Comment:
OK, I am confused by a lot of the logic of this test. I think I am seriously misunderstanding the ordering in which buffers are being spilled. Here is the sequence that I expect to happen.
- We have a pool of size 3N
- We allocate b1, b2, and b3, which fill up the pool
- We allocate b4, which triggers an OOM-spill event. Here is my first point of confusion. We implement LRU logic to determine what buffer to spill, so I would have expected b1 to be spilled at this point and for the memory to be ordered b4-b2-b3. You are then deleting b1 and b4 and claiming that frees up two fragments, but by my rationale that should only free up one because b1 should already have been evicted from device. What am I missing?

This leads to two branches. 

Branch 1:
Let us assume that you are correct and that `del b1, b4` actually deletes two buffers that are currently on device. That implies that b2 or b3 was spilled. Since buffers are ordered, I must conclude that b3 as the _most recently_ created buffer was what was spilled instead of the least recently created one, so we must now have our memory ordered b1-b2-b4 (not b4-b2-b3 as I originally surmised). If that is true, then the next `as_buffer` call should trigger an eviction of b2 and b4, because those are now the most recently created buffers. But that would not lead to any fragmentation, so this can't be right either.

Branch 2:
If my logic is correct and b1 was already evicted, then we should have b4-b2-b3 in memory and the next spill event should evict b2 and b3, memory which is then filled exactly by the new allocation and the assertion should fail. Since it's not, this also cannot be what is happening.

I'm a bit at a loss to understand exactly what _is_ happening. However, assuming all of the `assert` statements pass and everything works as expected, this code exposes another potential weak point of using the current measurement to determine fragmentation. Assuming that all the eviction ordering is happening in a sensible way, at the end we are allocating a buffer that triggers an OOM-spill event, but we aren't saving it to any local value, so the end result after `as_buffer` so it should be deallocated immediately, and as a result we're actually using 2N less memory than `lowest_utilization` claims.